### DISCUSSION TITLE
Sqlalchemy 2.0: any_ operator is failing mypy checks

### DISCUSSION THREAD

**Discussion by michaelbukachi** [sqlalchemy/sqlalchemy#9318] [Usage Questions] [answered]
What's the right syntax for using the `any_` operator so that it doesn't throw lint errors?
Mypy is complaining with the following snippet:
```python
from sqlalchemy import create_engine, select, Column, String, any_
from sqlalchemy.orm import sessionmaker, DeclarativeBase


class Base(DeclarativeBase):
    id = Column(String, primary_key=True)


class User(Base):
    phone = Column(String, nullable=False, index=True, unique=True)


if __name__ == '__main__':
    engine = create_engine("sqlite:///")
    Session = sessionmaker(engine)

    with Session.begin() as db:
        user = db.scalars(select(User.id).where(User.phone == any_(['1', '2', '3'])).limit(1)).first()
```
It's throwing the following error:
```bash
test.py:18: error: Argument 1 to "any_" has incompatible type "List[str]"; expected "Union[ColumnElement[<nothing>], _HasClauseElement, SQLCoreOperations[<nothing>], ExpressionElementRole[<nothing>], Callable[[], ColumnElement[<nothing>]], LambdaElement]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

**SQLAlchemy 2.0,3**

**Comment by zzzeek** [comment, marked as answer]
that's not the correct use of any_() , it doesnt accept a list of elements like that.  it seems to just go through in SQLite as the list bound to a single parameter,  but we dont have any notion of sqlite receiving lists as single parameters.

here's what your program returns if we actually add tablename and make it run:

```
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such function: ANY
[SQL: SELECT u.id 
FROM u 
WHERE u.phone = ANY (?)
 LIMIT ? OFFSET ?]
[parameters: (['1', '2', '3'], 1, 0)]
```

so I'd need to see the specific DB / operation you are looking to do

**Reply by zzzeek** [reply]
basically if you're on postgresql you would use the [array()](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.array) constructor

**Reply by michaelbukachi** [reply]
Oops. I forgot to change the engine to postgres. It's supposed to be:
`engine = create_engine("postgresql://test:test@localhost/test_db")`

Do you mean `any_(array(['1', '2', '3', '4']))`
?

**Reply by CaselIT** [reply]
that syntax seems from PostgreSQL.

You would either use it with `array()` or `values().scalar_values()`. Something like this
```py
import sqlalchemy as sa
from sqlalchemy.dialects import postgresql as pg

engine = sa.create_engine("postgresql://scott:tiger@/test", echo=True)

stmt = sa.select(
    sa.literal_column("'foo'") == sa.any_(pg.array(["1", "2", "3"]))
)
stmt2 =sa.select(
    sa.literal_column("'foo'")
    == sa.any_(
        sa.values(sa.column("x", sa.String))
        .data([("1",), ("2",), ("3",)])
        .scalar_values()
    )
)
print(stmt.compile(engine).string)
print(stmt2.compile(engine).string)

with engine.connect() as c:
    c.execute(stmt)
    c.execute(stmt2)
```

will print the expected
```
SELECT 'foo' = ANY (ARRAY[%(param_1)s, %(param_2)s, %(param_3)s]) AS anon_1
SELECT 'foo' = ANY (VALUES (%(param_1)s), (%(param_2)s), (%(param_3)s)) AS anon_1
2023-02-15 23:32:09,287 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-02-15 23:32:09,287 INFO sqlalchemy.engine.Engine SELECT 'foo' = ANY (ARRAY[%(param_1)s, %(param_2)s, %(param_3)s]) AS anon_1
2023-02-15 23:32:09,287 INFO sqlalchemy.engine.Engine [generated in 0.00024s] {'param_1': '1', 'param_2': '2', 'param_3': '3'}
2023-02-15 23:32:09,287 INFO sqlalchemy.engine.Engine SELECT 'foo' = ANY (VALUES (%(param_1)s), (%(param_2)s), (%(param_3)s)) AS anon_1
2023-02-15 23:32:09,287 INFO sqlalchemy.engine.Engine [no key 0.00012s] {'param_1': '1', 'param_2': '2', 'param_3': '3'}
2023-02-15 23:32:09,287 INFO sqlalchemy.engine.Engine ROLLBACK
```

**Reply by michaelbukachi** [reply]
Interesting. I didn't know about the `values().scalar_values()`. Thanks. 
In pre-2.0 version it used to work without having to explicitly pass an iterable into a postgres `array`, do you know why? :thinking:

**Reply by CaselIT** [reply]
>  I didn't know about the values().scalar_values(). Thanks.

that's new of v2 :)

> In pre-2.0 version it used to work without having to explicitly pass an iterable into a postgres `array`, do you know why? 🤔

You can still do that in v2, other than the type error the code correctly executes in postgresql.
Interestingly I though this was only psycopg2 that allowed it, but trying both psycopg3 and pg8000 both like that form.

@zzzeek any though on allowing it in the types?

**Reply by zzzeek** [reply]
if you are using a literal of values why not use ``in_()`` ?

@CaselIT what are we allowing in the types, plain list that goes through to the driver? we dont have a "list-in-place" type and starting to let something like that through without it being "legit" would be a bad idea.  asyncpg lets it throguh?

**Reply by CaselIT** [reply]
>  asyncpg lets it throguh?

yep
```
2023-02-16 00:04:58,781 INFO sqlalchemy.engine.Engine SELECT 'foo' = ANY ($1) AS anon_1
2023-02-16 00:04:58,781 INFO sqlalchemy.engine.Engine [generated in 0.00012s] (['1', '2', '3'],)
```

**Reply by michaelbukachi** [reply]
@zzzeek From some benchmarks I did, using `any_` produced more performant queries than `in_`. Especially when subqueries were involved.

**Reply by CaselIT** [reply]
PG usually rewriters in and not it in any/all but it doesn't always do that (according to the explain output at least).
But I've also noticed that in general any/all performs better thatn in/not_in

**Reply by LukasJerabek** [reply]
I think the problem is on sqlalchemy side. It is weird that they do not allow such syntax, you cant event find it in their documentation, but it is allowed in postgres and works as expected.

**Comment by LukasJerabek** [comment]
Actually I have just found a solution. You need to use it like this:
from sqlalchemy import any_, literal
column == any_(literal(['1', '2', '3']))

or even better:
from sqlalchemy import any_, literal, ARRAY

column == any_(literal(['1', '2'], ARRAY(String)))

**Reply by zoola969** [reply]
The second variant is more correct, because asyncpg fails on empty arrays if you use first variant.
